kyo-core/jvm/src/main/scala/kyo/StreamCompression.scala

```diff
@@ -0,0 +1,733 @@
+        ): Stream[Byte, Resource & IO & Ctx] =
+            deflateStream(stream.map(ev.apply), bufferSize, compressionLevel, strategy, flushMode, noWrap)
+        end deflate
+
+        inline def inflate(
+            bufferSize: Int = 1 << 15,
+            noWrap: Boolean = false
+        )(
+            using
+            tag: Tag[Emit[Chunk[Byte]]],
+            frame: Frame,
+            ev: A <:< Byte
+        ): Stream[Byte, IO & Resource & Ctx & Abort[StreamCompressionException]] =
+            inflateStream(stream.map(ev.apply), bufferSize, noWrap)
+        end inflate
+
+        inline def gzip(
+            bufferSize: Int = 1 << 15,
+            compressionLevel: CompressionLevel = CompressionLevel.Default,
+            strategy: CompressionStrategy = CompressionStrategy.Default,
+            flushMode: FlushMode = FlushMode.Default
+        )(
+            using
+            tag: Tag[Emit[Chunk[Byte]]],
+            frame: Frame,
+            ev: A <:< Byte
+        ): Stream[Byte, Resource & IO & Ctx] =
+            gzipStream(stream.map(ev.apply), bufferSize, compressionLevel, strategy, flushMode)
+        end gzip
+
+        inline def gunzip(
+            bufferSize: Int = 1 << 15
+        )(
+            using
+            tag: Tag[Emit[Chunk[Byte]]],
+            frame: Frame,
+            ev: A <:< Byte
+        ): Stream[Byte, IO & Resource & Ctx & Abort[StreamCompressionException]] =
+            gunzipStream(stream.map(ev.apply), bufferSize)
+        end gunzip
+    end extension
+
+    private def deflateStream[Ctx](
+        stream: Stream[Byte, Ctx],
+        bufferSize: Int,
+        compressionLevel: CompressionLevel,
+        strategy: CompressionStrategy,
+        flushMode: FlushMode,
+        noWrap: Boolean
+    )(
+        using
+        Tag[Emit[Chunk[Byte]]],
+        Frame
+    ) =
+
+        enum DeflateState derives CanEqual:
+            case Initialize                                                              extends DeflateState
+            case DeflateInput(delfater: Deflater, emit: Ack < (Emit[Chunk[Byte]] & Ctx)) extends DeflateState
+            case PullDeflater(
+                delfater: Deflater,
+                maybeEmitFn: Maybe[Ack => Ack < (Emit[Chunk[Byte]] & Ctx)],
+                chunk: Chunk[Byte]
+            ) extends DeflateState
+            case EmitDeflated(delfater: Deflater, maybeEmitFn: Maybe[Ack => Ack < (Emit[Chunk[Byte]] & Ctx)], chunk: Chunk[Byte])
+                extends DeflateState
+        end DeflateState
+
+        def emit(
+            stream: Stream[Byte, Ctx],
+            bufferSize: Int,
+            compressionLevel: CompressionLevel,
+            strategy: CompressionStrategy,
+            flushMode: FlushMode,
+            noWrap: Boolean
+        )(
+            using
+            Tag[Emit[Chunk[Byte]]],
+            Frame
+        ): Ack < (Resource & IO & Emit[Chunk[Byte]] & Ctx) =
+            val local = Local.init(new Array[Byte](bufferSize))
```

Comment:
I still don't understand why this can't be `val buffer = new Array[Byte](bufferSize)` and then use `buffer` directly in the loop. Also, an `IO` suspension is required for the initialization of the buffer since it's a mutable value.